vermouth/ffinput.py

```diff
@@ -51,196 +55,337 @@
+        section = self.section + [line.strip('[ ]').casefold()]
+        if tuple(section[-1:]) in self.METH_DICT:
+            self.section = section[-1:]
+        else:
+            while (tuple(section) not in self.METH_DICT
+                   and len(section) > 1):
+                ended.append(section.pop(-2))  # [a, b, c, d] -> [a, b, d]
+            self.section = section
+                          
+        result = None
+
+        if len(prev_section) != 0:
+            print(len(prev_section))
+            result = self.finalize_section(prev_section, ended)
+
+        action = self.header_actions.get(tuple(self.section))
+        if action:
+            action()
+
+        return result
+
+    def parse_section(self, line, lineno):
+        """
+        Parse `line` with line number `lineno` by looking up the section in
+        :attr:`METH_DICT` and calling that method.
+
+        On the contrary to the SectionLineParser, we can have wildcard sections.
+        Wildcard sections cannot have children sections.
+
+        Parameters
+        ----------
+        line: str
+        lineno: int
+
+        Returns
+        -------
+        object
+           The result returned by calling the registered method.
+        """
+        line = _substitute_macros(line, self.macros)
+        #wildcard_section = tuple(self.section[:-1] + ['*'])
+        end_section = []
+        if self.section:
+            end_section = self.section[-1]
+        if tuple(self.section) in self.METH_DICT:
+            method, kwargs = self.METH_DICT[tuple(self.section)]
+        #elif wildcard_section in self.METH_DICT:
+        #    method, kwargs = self.METH_DICT[wildcard_section]
+        else:
+            raise IOError("Can't parse line {} in section '{}' because the "
+                          "section is unknown".format(lineno, self.section))
+        try:
+            return method(self, line, lineno, **kwargs)
+        except Exception as error:
+            raise IOError("Problems parsing line {}. I think it should be a "
+                          "'{}' line, but I can't parse it as such."
+                          "".format(lineno, self.section)) from error
+
+    def finalize_section(self, pervious_section, ended_section):
+
+        """
+        Called once a section is finished. It appends the current_links list
+        to the links and update the block dictionary with current_block. Thereby it
+        finishes the reading a given section. 
+
+        Parameters
+        ---------
+        previous_section: list[str]
+            The last parsed section.
+        ended_section: list[str]
+            The sections that have been ended.
+        """
+
+        # type comparison is what makes this work!
+        
+        if type(self.current_block) != type(None):
+           self.blocks[self.current_block.name] = self.current_block
+
+        if type(self.current_link) != type(None):
+           self.links.append(self.current_link)
```

Comment:
@pckroon Because a block is a molecule and a molecule without atoms/graph evaluates to none. More precisely overwriting the __eq__ method in molecule has lead to the behavior that a molecule without atoms evaluates to None. This means unless you make a type comparison you cannot have a block without atoms.  If we want that is a completely different question. I don't know enough about the use case but it is test no. 1 in the test_ff_files and appears in other more integrated tests as well. Perhaps we could rais an IO error for these cases?

Reply:
Just to make sure we're on the same page, when you say "evaluates to none", do you mean:
a) `bool(mol) == bool(None) == False`
b) `mol == None`
c) `mol is None`

If there's tests allowing empty blocks/molecules I suggest we keep supporting that.

Reply:
a) but not c) and b) actually causes an attribute error for empty molecules

Reply:
I changed to the use of is.